### PR TITLE
Fix connect info serde impl and accepted string in cli commands

### LINKED ISSUE #1716: The connection string that setup UI generates is bunch of hex while `fedimint-cli` still expects ... some json?
I've just set up a regtest federation and the UI shows me QR code an base32 encoded blow, but `fedimint-cli join-federation` does not like it, and from I can tell still wants json.

Is there a way to convert between these?

@jkitman

### PR DESCRIPTION
**Pull request by Maan2003**
Fixes #1716

### CODE REVIEW

File: client/cli/src/main.rs [hunk 1]
```diff
@@ -326,12 +327,12 @@ trait ErrorHandler<T, E> {
         F: Fn(T) -> CliOutput;
 }
 
-impl<T, E: Error + 'static> ErrorHandler<T, E> for Result<T, E> {
+impl<T, E: Into<Box<dyn Error>>> ErrorHandler<T, E> for Result<T, E> {
```
Comment (Maan2003) [inline client/cli/src/main.rs:330]: because anyhow::Error does not impl Error

File: fedimint-core/Cargo.toml [hunk 2]
```diff
@@ -41,6 +41,7 @@ hbbft = { git = "https://github.com/jkitman/hbbft", branch = "upgrade-threshold-
 rand = "0.8.5"
 miniscript = { version = "7.0.0", git = "https://github.com/rust-bitcoin/rust-miniscript/", rev = "2f1535e470c75fad85dbad8633986aae36a89a92", features = [ "compiler", "serde" ] }
 secp256k1-zkp = { version = "0.7.0", features = [ "use-serde", "bitcoin_hashes", "global-context" ] }
+serde_with = "2.2.0"
```
Comment (mxz42) [inline fedimint-core/Cargo.toml:44]: that would be the introduction of a new dependency; though, I guess that'd be fine?

File: fedimint-server/src/config/io.rs [hunk 3]
```diff
@@ -154,6 +156,13 @@ fn plaintext_json_write<T: Serialize + DeserializeOwned>(
     Ok(())
 }
 
+fn plaintext_display_write<T: Display>(obj: &T, path: PathBuf) -> anyhow::Result<()> {
+    let mut file = fs::File::create(path.clone())
```
Comment (mxz42) [inline fedimint-server/src/config/io.rs:160]: nit: we can take reference here `&path`

File: scripts/cli-test.sh [hunk 4]
```diff
@@ -25,7 +25,7 @@ start_gateway
 #### BEGIN TESTS ####
 
 # test the fetching of client configs
-CONNECT_STRING=$(cat $FM_CFG_DIR/client-connect.json)
+CONNECT_STRING=$(cat $FM_CFG_DIR/client-connect)
```
Comment (mxz42) [inline scripts/cli-test.sh:28]: new ending to indicate filetype?  `.bech32` or similar?

**Review by mxz42 — COMMENTED** [review]
this change would also write the `client-connect.json` as bech32. 
previously you could open it and see something like: 

```json
{
  "urls": [
    "ws://127.0.0.1:18174/",
    "ws://127.0.0.1:18184/"
  ],
  "id": "82fa9f96e15668ef7a47034067632da29de736dc9388bb2aa21f562f76826fa7ee929af0ee28bd3324f01545cca0203a"
}
```

now the content would be

```
fed11kuncff50fcp3msef3aypv39jn78n405mw7gmal6uvpft0rw90dq95h8dgfhdzy2wuyx0gagqa5xpx9gqwaen5te0xyerwt3s9cczuvf6xyurzde59u2sqamn8ghj7vfjxuhrqt3s9ccn5vfcxyurgtcuz8av9
```

we're ok with that change?

File: fedimint-core/Cargo.toml (hunk 2 above)
Comment (Maan2003) [reply fedimint-core/Cargo.toml:44]: I can implement it manually, if we don't want to be have this dependency.

File: scripts/cli-test.sh (hunk 4 above)
Comment (Maan2003) [reply scripts/cli-test.sh:28]: IMO extensionless is fine. json is a well know format support by many editors, bech32 is not.

File: fedimint-core/src/api.rs [hunk 5]
```diff
@@ -1099,7 +1100,11 @@ mod tests {
         };
 
         let bech32 = connect.to_string();
+        let json = serde_json::to_string(&connect).unwrap();
+        assert_eq!(json, format!("\"{bech32}\""));
```
Comment (dpc) [inline fedimint-core/src/api.rs:1104]: json and bech32 being interwined just make reading harder.

I'm fine with one test, but first render and compare bech32, then render and compare json (or the other way around). Just not two at the same time.

**Comment by justinmoon** [comment]
> we're ok with that change?

Yes. This is our new connection string format. Is there an easy way to convert this bech32 back into JSON in CLI for debugging?

Tried `hal` but I just get hex:

```
% hal bech32 decode fed11kuncff50fcp3msef3aypv39jn78n405mw7gmal6uvpft0rw90dq95h8dgfhdzy2wuyx0gagqa5xpx9gqwaen5te0xyerwt3s9cczuvf6xyurzde59u2sqamn8ghj7vfjxuhrqt3s9ccn5vfcxyurgtcuz8av9
{
  "bech32": "fed11kuncff50fcp3msef3aypv39jn78n405mw7gmal6uvpft0rw90dq95h8dgfhdzy2wuyx0gagqa5xpx9gqwaen5te0xyerwt3s9cczuvf6xyurzde59u2sqamn8ghj7vfjxuhrqt3s9ccn5vfcxyurgtcuz8av9",
  "hrp": "fed1",
  "payload": "161c13180909140f091801111b101909111d04010c110512131e0713150f141b0e1e081b1d1f1a1c0c01090b0f030e050f0d00051417070d0809170d02040a0e1c04060f081d08001d140601060508000e1d1913140b190f060419030e0b1110051818021c0c091a06041c03020d1914051c0a10001d1b13070817121e0c0912061c1703000b111005181813140c091806041c03080b18"
}
```

File: fedimint-core/src/api.rs (hunk 5 above)
Comment (dpc) [inline fedimint-core/src/api.rs:1104]: Wait. I don't get it. What is going on here? json formatting is `\"{bech32}\"`? How does this ... what?

File: Cargo.lock [hunk 6]
```diff
@@ -733,6 +765,85 @@ dependencies = [
...
+name = "cxx"
+version = "1.0.91"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "86d3488e7665a7a483b57e25bdd90d0aeb2bc7608c8d0346acf2ad3f1caf1d62"
+dependencies = [
+ "cc",
+ "cxxbridge-flags",
+ "cxxbridge-macro",
+ "link-cplusplus",
+]
```
Comment (justinmoon) [inline Cargo.lock:778]: I suspect this was from another change. Could you update `Cargo.lock` to only include changes with `serde_with`?

**Comment by dpc** [comment]
> @dpc would this require updating the API versions?

I don't think so. The original change was, but it was landed around the same time we landed versioning.

File: fedimint-core/src/api.rs (hunk 5 above)
Comment (Maan2003) [reply fedimint-core/src/api.rs:1104]: \ is just escape here. so format is `"{bech32}"`

File: Cargo.lock (hunk 6 above)
Comment (Maan2003) [reply Cargo.lock:778]: I am not sure how this is coming in. But this change still stays when regenerating cargo.lock

Comment (Maan2003) [reply Cargo.lock:778]: I think this is due to optional deps of serde_with, just writing manually impl.

File: fedimint-core/src/api.rs (hunk 5 above)
Comment (dpc) [reply fedimint-core/src/api.rs:1104]: Thank you. I need more coffee.

**Review by dpc — APPROVED** [review]

File: scripts/cli-test.sh (hunk 4 above)
Comment (mxz42) [reply scripts/cli-test.sh:28]: yah, I agree. 
My thinking was more in regards to indicate what the numbers and letters are in this file. Putting `.bech32` at the end means someone does not have to look at (perhaps lacking/outdated) documentation or figure out what it is

I'm fine with either way, though

**Review by dpc — APPROVED** [review]

**Review by elsirion — APPROVED** [review]

**Pull request merged**